src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/HandlerProvider.cs

```diff
@@ -164,19 +202,20 @@ static void VerifyHandlers(IEnumerable<RequestHandlerMetadata> requestHandlerKey
 
     private static readonly IReadOnlyList<string> RequiredMethods = new List<string> { "initialize", "initialized", "shutdown", "exit" };
 
+    private record HandlerTypes(Type? RequestType, Type? ResponseType, Type RequestContext);
+
     /// <summary>
     /// Retrieves the generic argument information from the request handler type without instantiating it.
     /// </summary>
-    private static (Type? requestType, Type? responseType, Type requestContext) ConvertHandlerTypeToRequestResponseTypes(Type handlerType)
+    private static IEnumerable<HandlerTypes> ConvertHandlerTypeToRequestResponseTypes(Type handlerType)
     {
-        var requestHandlerGenericType = handlerType.GetInterfaces().Where(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(IRequestHandler<,,>)).SingleOrDefault();
-        var parameterlessNotificationHandlerGenericType = handlerType.GetInterfaces().Where(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(INotificationHandler<>)).SingleOrDefault();
-        var notificationHandlerGenericType = handlerType.GetInterfaces().Where(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(INotificationHandler<,>)).SingleOrDefault();
-
-        Type? requestType;
-        Type? responseType;
-        Type requestContext;
-        if (requestHandlerGenericType is not null)
+        var requestHandlerGenericTypes = handlerType.GetInterfaces().Where(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(IRequestHandler<,,>));
```

Comment:
>handlerType.GetInterfaces() [](http://example.com/codeflow?start=41&length=27)

Might be nice to put this in a local and reuse it as it doesn't look to be free

Reply:
This looks nicer, but it still does "GetInterfaces().Where(i => i.IsGenericType" three times, no?

Reply:
Ah, I misunderstood your original comment, updated.